com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java

```diff
@@ -65,35 +72,82 @@ static public GeneratedMessageV3.Builder<?> newBuilder(String ext) throws Compil
         }
     }
 
-    protected B transform(Task<Void> task, IResource resource, B messageBuilder) throws IOException, CompileExceptionError {
+    protected B transform(Task task, IResource resource, B messageBuilder) throws IOException, CompileExceptionError {
         return messageBuilder;
     }
 
+    /**
+     * Scan proto message and create a sub-task for each resource in it
+     * @param builder message or builder of the file that should be scanned
+     * @param taskBuilder the builder where result should be applied to
+     */
+    protected void createSubTasks(MessageOrBuilder builder, Task.TaskBuilder taskBuilder) throws CompileExceptionError {
+        List<Descriptors.FieldDescriptor> fields = builder.getDescriptorForType().getFields();
+        for (Descriptors.FieldDescriptor fieldDescriptor : fields) {
+            DescriptorProtos.FieldOptions options = fieldDescriptor.getOptions();
+            Descriptors.FieldDescriptor resourceDesc = DdfExtensions.resource.getDescriptor();
+            boolean isResource = (Boolean) options.getField(resourceDesc);
+            Object value = builder.getField(fieldDescriptor);
+            if (value instanceof List) {
+                @SuppressWarnings("unchecked")
+                List<Object> list = (List<Object>) value;
+                for (Object v : list) {
+                    if (isResource && v instanceof String) {
+                        createSubTask((String) v, fieldDescriptor.getName(), taskBuilder);
+                    } else if (v instanceof MessageOrBuilder) {
+                        createSubTasks((MessageOrBuilder) v, taskBuilder);
+                    }
+                }
+            } else if (isResource && value instanceof String) {
+                boolean isOptional = fieldDescriptor.isOptional();
+                String resValue =  (String) value;
+                // We don't require optional fields to be filled
+                // if such a field has no value - just ignore it
+                if (isOptional && resValue.isEmpty()) {
+                    continue;
+                }
+                createSubTask(resValue, fieldDescriptor.getName(), taskBuilder);
+            } else if (value instanceof MessageOrBuilder) {
+                createSubTasks((MessageOrBuilder) value, taskBuilder);
+            }
+        }
+    }
+
+    // This used to parse the main input resource ('firstInput()' or 'input.get(0)') and then reuse it on all the stages.
+    protected B getSrcBuilder(IResource input) throws IOException, CompileExceptionError {
```

Comment:
Now we parse every proto-resource on create stage to be able to read options for each field and then create tasks. 
This function helps to parse input resource once and cache it to be able to reuse it later on build stage and/or some other places.
Before that, we did it 2 and sometimes even 3 times per task.